Feat: custom color theme

gambhirsharma opened the pull request:
closes #1280

Selected color applies to all primary UI elements and syncs across main window and cluster iframes.

## Changes 

- **Accent color selector** with visual swatches in theme preferences
- **Reset to Default** button (shown when non-default selected)
- **Custom accent color preference** with persistence
- **Theme computation** overriding primary colors (blue, primary, buttonPrimaryBackground, menuActiveBackground, etc.)
- **IPC-based theme sync** between root frame and cluster iframes
- **Theme persistence layer** for cluster frames

## Architecture

1. **Preference Storage** - Stores/retrieves accent color from user preferences with persistence
2. **UI Components** - Color picker in Preferences panel with visual swatches
3. **Theme Computation** - Merges base theme with accent color overrides (primary, blue, buttonPrimaryBackground, etc.)
4. **Theme Application** - Applies CSS custom properties to DOM (`--primary`, `--blue`, etc.)
5. **Theme Synchronization** - IPC messaging system for cross-frame sync

**Key Design Decisions:**
- **IPC channels** for cross-origin iframe communication (cluster frames use different subdomains)
- **Multiple broadcast retries** (50ms, 150ms, 300ms) to handle async Dependency Injection container initialization
- **Global backup storage** (`window.__lastReceivedTheme`) to persist theme across document reloads and CSS variable resets

### Screenshots
<img width="1440" height="900" alt="image" src="https://github.com/user-attachments/assets/9ef90b17-a485-4512-8b1c-f2c3312ba935" />


<details>
<summary>view more screen-shot</summary>
<img width="1440" height="900" alt="image" src="https://github.com/user-attachments/assets/bed95386-a362-4e95-8406-6de91c037156" />

<img width="1440" height="900" alt="image" src="https://github.com/user-attachments/assets/b7bef350-2753-42ca-bcea-54a2cf024e62" />
</details>

## Testing

- [x] Color picker in Preferences 
- [x] Persists after restart 
- [x] Syncs to iframes 
- [x] Reset button 
- [x] Light/Dark themes

## Preset Colors

- **Teal** (#00a7a0) - Default
- **Green** (#4caf50)
- **Blue** (#2196f3)
- **Orange** (#ff9800)

gambhirsharma committed:
feat: custom color theme

gambhirsharma committed:
chore: Format code biome

robertobandini commented:
> Hi @ robertobandini, sorry for creating this PR so late 🥲

Hi @gambhirsharma no prob, on the contrary thank you very much!

> * Should the Freelens logo colour also change according to the theme it has applied?

No, I think the logo must remain as originally designed.

> * For now, [this feature](https://github.com/freelensapp/freelens/issues/550#issuecomment-2824022298) is not being implemented on this MR do we need this?

No, if the user can choose the color from the UI using your feature.

I tought you had to review [this PR](https://github.com/freelensapp/freelens/pull/1368) but @dex4er did it before you, so I suppose you opened this PR to propose your solutions.
I haven't tried it yet, but seems good, thanks!

gambhirsharma committed:
chore: Gambhir Sharma added to ADOPTERS list

robertobandini commented:
Thanks @omarluq and @gambhirsharma!
We have time as we are planning to include this feature on the 1.8.0, so we can do the poll.
@gambhirsharma if you can maybe it is better to creare the poll in the discussion here in the repo so you can share the link on Discord and everybody can vote.

robertobandini commented:
> We could add both a list of preset colors and a color picker, or at least a section where you can enter the color code. What do you think about adding an option in the dropdown that shows the colored square with some different colors and a "custom" label, and on click it opens the picker?

I admit that I tought the same, could be the third option on the poll, what do you think guys?

gambhirsharma committed:
chore: update snapshot

gambhirsharma committed:
chore: update snapshot

gambhirsharma commented:
> I admit that I tought the same, could be the third option on the poll, what do you think guys?

Yes, that's a great idea. I will implement it. 

For the poll, @robertobandini, I wasn’t sure which colors to include as options in the GitHub discussion. I posted a message on Discord to start the conversation, and once we see which colors are most popular, we can add them to the presets.

Also, @robertobandini, could you review this MR or assign someone to take a look at it

robertobandini commented:
@mariomamo If you would like you can review this too, just let know, thanks!

mariomamo commented on packages/core/src/features/pod-logs/__snapshots__/download-logs.test.tsx.snap:
Why this snap has been deleted?

gambhirsharma replied:
Sorry, I wasn’t able to update the snapshots after the new UI changes. So, in my confusion, I thought deleting the existing snapshots and regenerating them from the updated UI might be the right approach.

Could you please explain how snapshot testing works and what the correct way is to update snapshots when the UI changes?

```bash
# Delete all snapshots first
find . -type d -name __snapshots__ -exec rm -rf {} + 2>/dev/null

# Then run tests with update flag
pnpm test:unit:updatesnapshot
```

omarluq replied:
hey @gambhirsharma to update a snap
- run `cd packages/{package-where-snap-is}`
- run `CI=false pnpm test:unit -- -u` to run all tests and update snaps or alternatively you can run `pnpm jest <testfile> -u`
- after that just stage and commit the updated snap file

lmk if you face any issues with this 👍🏼

gambhirsharma committed:
fix: snapshot update

gambhirsharma committed:
Merge branch 'revert-feat/theme' into feat/theme

Revert the updated snapshot & update the snapshot properly

MatiasRoje committed:
Merge branch 'main' into feat/theme

MatiasRoje committed:
register theme-related injectables via buid:di

omarluq commented on packages/core/src/features/welcome/__snapshots__/navigation-using-application-menu.test.ts.snap:
```diff
@@ -549,81 +549,171 @@ exports[`welcome - navigation using application menu when navigated somewhere el
+                                <path
+                                  d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
+                                />
+                              </svg>
```
can we move this to `packages/ui-components/icon/assets/{name-svg}.svg‎` instead on inline plz

gambhirsharma commented:
> @gambhirsharma Is the DI syncing process not working out of the box for this feature? Did you find some problem? An alternative to the current implementation would be to have userPreferences state to emit some "themeChanged" event and let the clusterFrames and UI elements react to it by themselves.G

hi, @MatiasRoje, DI syncing wasn't working for me. I got stuck on it, so I decided to implement this multi-tries & global backup storage. Initially, I only wanted the feature to work & didn't think much about anti-pattern.

I will try to look into the project again and adopt a better approach. 

Sorry for the delays just completed the exams today, so I will try to complete this PR ASAP

gambhirsharma committed:
refactor: simplify theme system

- Replace pushed theme updates with notification-based pattern
- Each frame now derives its active theme from local state
- Remove iframe theme broadcasting and copy-theme-from-parent logic
- Improve accent color override to only replace colors using default
accent
- Move color swatch styles to CSS module

gambhirsharma committed:
chore: update snapshot

gambhirsharma committed:
restore deleted snaps

gambhirsharma committed:
chore: update snapshot after restoring

MatiasRoje commented:
Hey @gambhirsharma, good progress so far. We want to get this feature merged soon, so let's keep working on it.

A few blocking points to address before we merge:

- Please remove the separate theme IPC/update path (`themeChangedChannel` and related listener/wiring). User preferences storage already syncs state across frames, so this creates a duplicate propagation path for the same data and is not consistent with the rest of the codebase.
- Make accent overrides explicit by key while iterating through the baseTheme.color objects. Matching the theme color override by raw hex value (`value === #00a7a0`) is a bit fragile and could have side effects for future themes or extensions. It would be better to define a fixed list beforehand and use that for the matching: 
const accentKeys: LensColorName[] = [
    "primary",
    "menuActiveBackground",
    "buttonPrimaryBackground",
    // ...
  ];
- Add a couple of reliable tests for this feature.

A note on the accent colors:

- Blue looks fine. Green and orange are already being used as the colorOk and colorWarning in the light/dark themes. I do like them as options but we should go for different shades or tints that are unique and appropriate for a primary color.
Please consider good readability over white and dark backgrounds.

The feature is close and on the right track. Once these points are addressed I think we are ready to merge. I would add a couple of more nice-looking color options and don't introduce the color-picker in this PR. I would focus on what we have already and improve it over time if the community requests it. Would you agree with that @robertobandini ?

gambhirsharma commented:
Thanks @MatiasRoje, for the review! I will push the changes ASAP

> The feature is close and on the right track. Once these points are addressed I think we are ready to merge. I would add a couple of more nice-looking color options and don't introduce the color-picker in this PR. I would focus on what we have already and improve it over time if the community requests it. Would you agree with that @robertobandini ?

I believe this would be a better approach to go with for now. I have implemented this color picker in my local branch, but I don't feel confident with the implementation; that's the reason why I didn't add the commits in this PR. Let me know what you think @robertobandini, about it.

robertobandini commented:
Hi guys @gambhirsharma @MatiasRoje, thanks!
I agree with you; it's best to finalize it now, as Matias also said, and then further improve it based on community requests.

gambhirsharma committed:
refactor: theme IPC/update themeChangedChannel

gambhirsharma committed:
refactor: use explicit accent key list for custom accent override

gambhirsharma committed:
test: cover active theme logic with unit tests

gambhirsharma committed:
chore: add more accentColorOptions

dex4er committed:
Merge branch 'main' into feat/theme

gambhirsharma committed:
chore: sort imports alphabetically in active-theme test

gambhirsharma committed:
Merge branch 'freelensapp:main' into main

gambhirsharma committed:
refactor: extract accent colors to shared module

- Move DEFAULT_ACCENT_COLOR and accentColorOptions to accent-colors.ts
- Fix active theme to preserve third-party sidebarActiveColor when no
accent keys match (prevents incorrect overrides)
- Add test case for themes using custom accent colors

gambhirsharma committed:
Merge branch 'freelensapp:main' into main

gambhirsharma committed:
Merge branch 'main' into feat/theme

gambhirsharma committed:
update: snapshot

gambhirsharma committed:
fix: add aria-label to theme and accent color selects

gambhirsharma committed:
chore: restore &  upgrade snapshot
